testutil-server/src/main/java/io/spine/testing/server/blackbox/FailedHandlerGuard.java

```diff
@@ -0,0 +1,113 @@
+ */
+
+package io.spine.testing.server.blackbox;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import io.spine.server.event.AbstractEventSubscriber;
+import io.spine.server.type.EventClass;
+import io.spine.server.type.EventEnvelope;
+import io.spine.system.server.HandlerFailedUnexpectedly;
+
+import static io.spine.json.Json.toJson;
+import static java.lang.String.format;
+import static java.lang.System.lineSeparator;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/**
+ * Performs logging of failed signals handlers or fails the test depending the current
+ * context exception tolerance.
+ */
+final class FailedHandlerGuard extends AbstractEventSubscriber implements DiagnosticLogging {
+
+    private HandlerFailureTolerance tolerance = HandlerFailureTolerance.RAISE;
+
+    @Override
+    public ImmutableSet<EventClass> messageClasses() {
+        return EventClass.setOf(HandlerFailedUnexpectedly.class);
+    }
+
+    /**
+     * Always dispatched {@link HandlerFailedUnexpectedly} event.
+     */
+    @Override
+    public boolean canDispatch(EventEnvelope eventEnvelope) {
+        return true;
+    }
+
+    @Override
+    protected void handle(EventEnvelope eventEnvelope) {
+        HandlerFailedUnexpectedly event = (HandlerFailedUnexpectedly) eventEnvelope.message();
+        on(event);
+    }
+
+    /**
+     * Throws an {@link org.opentest4j.AssertionFailedError AssertionFailedError}
+     * unless the guard is configured to {@linkplain #tolerateFailures() tolerate} exceptions.
+     *
+     * <p>If the guard is tolerating exceptions, logs the handler failure.
+     */
+    @VisibleForTesting
+    void on(HandlerFailedUnexpectedly event) {
+        String msg = format(
+                "The entity (state type `%s`) could not handle the signal `%s`:%n%s%n",
+                event.getEntity().getTypeUrl(),
+                event.getHandledSignal().getTypeUrl(),
+                event.getError().getMessage()
+        );
+        switch (tolerance) {
+            case LOG:
+                log(msg, event);
+                break;
+            case RAISE:
+            default:
+                fail(() -> msg + lineSeparator() + toJson(event));
+                break;
+        }
+    }
+
+    /**
+     * Asks the guard to tolerate exceptions.
+     */
+    void tolerateFailures() {
+        tolerance = HandlerFailureTolerance.LOG;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>The {@code BlackBoxBoundedContext} only consumes domestic events.
+     */
+    @Override
+    public ImmutableSet<EventClass> domesticEventClasses() {
+        return eventClasses();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>The {@code BlackBoxBoundedContext} does not consume external events.
+     */
+    @Override
+    public ImmutableSet<EventClass> externalEventClasses() {
+        return ImmutableSet.of();
```

Comment:
How about introducing  `EventClass.emptySet()` so that it's easier to find places where return empty sets?